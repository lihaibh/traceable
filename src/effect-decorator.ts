import 'reflect-metadata';

import { EFFECTS_META_KEY, SUBSCRIPTIONS_META_KEY, TRACEABLE_META_HOOK } from '@lib/meta';
import { chain, exists, has, isNotNil, isObservable, isPrototypeOf, prop } from '@lib/utils/common';
import * as _ from 'lodash';
import * as R from 'ramda';
import { Observable } from 'rxjs/Observable';
import { Subscription } from 'rxjs/Subscription';
import { noop } from 'rxjs/util/noop';
import { isArray } from 'util';

/**
 * Register a side effect to the prototype in order to run it in a traceable object.
 *
 * @returns {PropertyDecorator}
 */
export function Effect(): PropertyDecorator {
    return (target, effect: string): void => {
        // According the docs, target wil be a constructor for static members or prototype for instance members
        const original_target = target;
        const original_constructor = _getConstructor(target);
        const effect_is_static = _isConstructor(target);

        // Validate whether we are in a traceable class type
        if (!_isTraceable(original_target)) {
            throw new TypeError('You can\'t use @Effect decorator in a non traceable prototype');
        }

        // Validate Effect
        if (!_isEffectObservable(original_target, effect)) {
            throw new TypeError(`Side effect ${original_constructor.name}.${effect} is not an Observable`);
        }

        // Add the effect as a key to the target
        _addEffect(original_target, effect);

        if (effect_is_static) {
            // Run static effect
            _addSubscription(
                // Static members should be invoked as soon as they are initialized
                original_target,
                // Create a subscription (which activates the effect instantly)
                prop(effect as never, original_constructor).subscribe(noop),
            );
        }
    };
}

/**
 * Stop all effect subscriptions of a class and its base prototypes
 *
 * @param constructor
 */
export function STOP_ALL_EFFECTS(constructor_derived: { new(...args: any[]): any }) {
    _.each(_getConstructorChainDesc.call(constructor_derived), (constructor_base) => {
        STOP_OWN_EFFECTS(constructor_base);
    });
}

/**
 * Stops only static effects subscriptions of this own class.
 *
 * @param constructor
 */
export function STOP_OWN_EFFECTS(constructor: new (...args: any[]) => any) {
    if (!_isTraceable(constructor)) {
        throw new TypeError(`You can't stop static effects in a non traceable type: ${constructor.name}`);
    } else {
        const own_subscriptions = prop(SUBSCRIPTIONS_META_KEY as never, constructor);

        if (isArray(own_subscriptions)) {
            _.each(own_subscriptions, (subscription) => {
                if (isPrototypeOf(Subscription, subscription)) {
                    subscription.unsubscribe();
                }
            });

            _.remove(own_subscriptions, isPrototypeOf(Subscription));
        }
    }
}

function _isEffectObservable(target, effect): boolean {
    // Check the property type is in the right type
    const type = Reflect.getMetadata('design:type', target, effect);
    return (type === Observable.prototype.constructor) ||
        isObservable(prop(effect, target));
}

function _isTraceable(target: any): boolean {
    const constructor = _getConstructor(target);
    const prototype = _getPrototype(target);

    if (constructor.name === 'StateTraceable') {
        return true;
    } else {
        // In order to avoid circular dependencies - require the module on runtime
        const StateTraceable = require('@app/core/traceable/state-traceable').StateTraceable;
        return isPrototypeOf(StateTraceable, prototype) ||
            Reflect.hasMetadata(TRACEABLE_META_HOOK, constructor);
    }
}

function _getConstructor(target): { new(...args: any[]): any } {
    return _isConstructor(target) ? target : prop('constructor', target);
}

function _getPrototype(target): any {
    return _isConstructor(target) ? prop('prototype', target) : target;
}

export function getEffects() {
    let takeEffects = _takeInstanceEffects;

    if (_isConstructor(this)) {
        takeEffects = _takeStaticEffects;
    }

    // get only those in the prototype chain and order them by inheritance (super class first)
    return _.chain(_getConstructorChainDesc.call(this)).map(takeEffects).flattenDeep().filter(isNotNil).value() as string[];
}

function _isConstructor(target): boolean {
    return has('prototype', target);
}

export function getSubscriptions() {
    // get only those in the prototype chain and order them by inheritance (super class first)
    return _getConstructorChainDesc.call(this).map(_takeSubscription).filter(isPrototypeOf(Subscription)) as Subscription[];
}

function _takeSubscription(constructor: { new(...args: any[]): any }) {
    return prop(SUBSCRIPTIONS_META_KEY as never, constructor);
}

function _takeInstanceEffects(constructor: { new(...args: any[]): any }) {
    return prop(EFFECTS_META_KEY as never, constructor.prototype);
}

function _takeStaticEffects(constructor: { new(...args: any[]): any }) {
    return prop(EFFECTS_META_KEY as never, constructor);
}

export function START_OWN_EFFECTS(constructor: { new(...args: any[]): any }) {
    // TODO: what if we already subscribed?
}

export function START_ALL_EFFECTS(constructor: { new(...args: any[]): any }) {
    // TODO: what if we already subscribed
}

function _getConstructorChainDesc() {
    return _.reverse(_.uniqBy(chain.call(this), prop('name')));
}

const _addTargetMetaData = (meta_key: string, target: any, data: any) => {
    // Add data directly to the target (constructor or prototype)
    const target_own_data = target.hasOwnProperty(meta_key) ?
        target[meta_key] : _.set(target, meta_key, [])[meta_key];

    if (exists(data, target_own_data)) {
        throw new Error(`Trying to insert duplicated values to: ${meta_key} in target: ${_getConstructor(target).name}`);
    } else {
        target_own_data.push(data);
    }
};

const _addEffect = R.curry(_addTargetMetaData)(EFFECTS_META_KEY);
const _addSubscription = R.curry(_addTargetMetaData)(SUBSCRIPTIONS_META_KEY);
