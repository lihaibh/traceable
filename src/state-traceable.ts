import 'rxjs/add/operator/distinctUntilChanged';
import 'rxjs/add/operator/do';
import 'rxjs/add/operator/filter';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/multicast';
import 'rxjs/add/operator/take';

import { Action } from '@lib/contracts/action';
import { Location } from '@lib/contracts/location';
import { Traceable } from '@lib/contracts/traceable';
import { Transformer } from '@lib/contracts/transformer';
import { Effect, getEffects } from '@lib/effect-decorator';
import { equals, isFunction, isObservable, path } from '@lib/utils/common';
import { selectOperator as select } from '@lib/utils/rx-operators/select';
import * as _ from 'lodash';
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
import { Observable } from 'rxjs/Observable';
import { ReplaySubject } from 'rxjs/ReplaySubject';
import { Subscription } from 'rxjs/Subscription';
import { noop } from 'rxjs/util/noop';
import { Memoize } from 'typescript-memoize';
import { isArray, isString } from 'util';

export abstract class StateTraceable<STATE, ACTIONS extends Action> implements Traceable<STATE, ACTIONS> {
    /**
     * Start running side effects in the background from this class and sub-classes.
     */
    public start() {
        if (!this._initialized) {
            this.dispatch(this.initialize());
            this._initialized = true;
        }

        // Get effect property names
        const effects = getEffects.call(this);

        // subscribe all side effects
        this._side_effects = _.map(effects, (effect: string) => {
            const stream = this[effect];

            // subscribe to them
            if (isObservable(stream)) {
                return stream.subscribe(noop);
            } else {
                throw new Error(`${effect} is not an observable`);
            }
        });
    }

    /**
     * Dispatching a new state to listeners.
     *
     * @param {STATE} state - new state to dispatch
     */
    protected dispatch(state: STATE) {
        this._state_dispatcher.next(state);
    }

    /**
     * Feeding the dispatcher with an initialized state.
     * This is the first state we will emit to subscribers.
     *
     * @returns {STATE} the first state
     */
    protected abstract initialize(): STATE;

    /**
     * Get updates of new state changes as a stream.
     *
     * @param {string | string[]} location
     * @param {string} rest
     * @returns {Observable<S | STATE>}
     */
    @Memoize()
    public state$<S>(location?: Location<STATE, S>, ...rest: string[]): Observable<S | STATE> {
        const source$ = this._source$();

        if (isString(location)) {
            // TODO: Check if it has dots in it
            return source$.pipe(select(location, ...rest));
        } else if (isArray(location) || isFunction(location)) {
            return source$.pipe(select.apply(source$, location));
        }

        return source$;

        // when subscribing to data in any hierarchy
        // TODO: we might want to perform some actions to activate side effects before selecting data

        // TODO: support also selection by function by decorating the state that we pass to this function
        // with a property descriptor to track the path to each of the elements
    }

    @Memoize()
    private _source$(): Observable<STATE> {
        return this._state_dispatcher.asObservable()
            /* We need this line in order to compare by content rather than by reference
            * basically the dispatcher can dispatch same object reference with updated data */
            .map(state => _.cloneDeep(state))
            .distinctUntilChanged(equals)
            .multicast(new ReplaySubject(1)).refCount() as Observable<STATE>;
    }

    /**
     * Stops all side effects from running in the background.
     * it also stops default side effects that comes with this class.
     */
    public stop() {
        _.each(this._side_effects, (subscription: Subscription) => subscription.unsubscribe());
    }

    /**
     * Get the last state by selection.
     * We are getting the last emitted state
     *
     * @param {Function | string[]} location
     * @returns {S | STATE} the data located inside the current state by the path or the whole state
     */
    public snapshot<S>(location?: Location<STATE, S>, ...rest: string[]): STATE | S {
        let snapshot: STATE | S = this._state_dispatcher.getValue();

        if (isFunction(location)) {
            snapshot = (location as Transformer<STATE, S>)(snapshot);
        } else if (isString(location)) {
            snapshot = path(location, snapshot);
        }

        return snapshot as S | STATE;
    }

    /**
     * Get actions performed on this class as a stream.
     *
     * @returns {Observable<ACTIONS extends Action>}
     */
    @Memoize()
    public action$(): Observable<Action> {
        return this._actions_dispatcher.asObservable()
            .multicast(new ReplaySubject(1)).refCount();
    }

    /**
     * Perform a new action to trigger a side effect or to change the state.
     *
     * @param {ACTIONS} action
     */
    public act(action: ACTIONS): void {
        this._actions_dispatcher.next(action);
    }

    /**
     * Gets the action and the current state to compute a new state to dispatch.
     *
     * @param {ACTIONS} action - current action
     * @param {STATE} state
     * @see act
     */
    protected abstract reduce(action: ACTIONS, state: STATE): STATE;

    /**
     * Manage the actions to operate side effects and change state with new values.
     */
    protected _actions_dispatcher: ReplaySubject<ACTIONS> = new ReplaySubject(1);
    /*
    * We must have these data members initialized beforehand
    * so that we can declare side effects that will be initialized on class declaration time
    */
    protected _state_dispatcher: BehaviorSubject<STATE> = new BehaviorSubject<STATE>(null);

    /**
     * A side effect that binds the actions to the reducer function in order to dispatch state updates.
     * When a new action comes in, we execute the reducer function to compute the next state that will multicast
     * to listeners.
     * Other side effects from the subclasses might trigger an action to change the state or we can do it directly
     * by executing an action.
     */
    @Effect()
    private _state_update: Observable<any> = this.action$()
        // combine action with the current state
        // it will always generate new object so distinct function will compare by content rather than by reference
        .map(__ => [__, this.snapshot()])
        // spare calls to the reduce function if combination of action and state did not result with a new state
        .distinctUntilChanged(equals)
        /*
        * run reducer in order to calculate the new state by the given action.
        *
        * clone the state instance - to compare against the previous state.
        * because the reducer may not result with new instance, rather it modifies the state argument
        * without this line operator 'distinctUntilChanged' might fail detect changes.
        */
        .map(__ => this.reduce.apply(this, _.cloneDeep(__)))
        // deep scan for state changes against the previous
        .distinctUntilChanged(equals)
        // notify all state listeners about the change
        .do(this.dispatch.bind(this));

    /**
     * We basically handle all instance' side-effects subscriptions here.
     */
    private _side_effects: Subscription[];

    // Determine if the object was initialized
    private _initialized = false;
}
