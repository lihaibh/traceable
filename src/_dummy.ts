import 'rxjs/add/observable/from';
import 'rxjs/add/operator/switchMapTo';

import { Effect } from '@lib/effect-decorator';
import ofType from '@lib/rx-operators/oftype';
import { StateTraceable } from '@lib/state-traceable';
import { Action } from '@lib/types';
import * as _ from 'lodash';
import { Observable } from 'rxjs/Observable';

export const first_state = {
    lvl0: {
        data: 0,
        lvl1: {
            data: 1,
        },
    },
};

export class Dummy extends StateTraceable<DummyState, DummyActions> {
    constructor(private _init_state = first_state) {
        super();
        this.start();
    }

    public manualChangeLvl0(change_to: number) {
        this.act(new ChangeLvl0(change_to));
    }

    public triggerEffectForStateAutoUpdate() {
        this.act(new AutoUpdateLvl1());
    }

    protected reduce(action: DummyActions, state: DummyState): DummyState {
        // We must create new object instance because distinctUntilChanged might compare by references
        if (action.type === 'change-lvl0') {
            return _.set(state, 'lvl0.data', (action as ChangeLvl0).payload);
        } else if (action.type === 'change-lvl1') {
            return _.set(state, 'lvl0.lvl1.data', (action as ChangeLvl1).payload);
        }

        return state;
    }

    protected initialize(): DummyState {
        return this._init_state;
    }

    @Effect()
    private _update_lvl1_data_action: Observable<any> =
        this.action$().pipe(ofType('auto-update'))
            .switchMapTo(
                Observable.from(fibonacci() as any).take(10)
                    .map((x: number) => new ChangeLvl1(x)),
        )
            // Invoke the action
            .do(this.act.bind(this));
}

// tslint:disable-next-line:max-classes-per-file
export class ChangeLvl0 implements Action {
    type = 'change-lvl0';

    constructor(public payload: number) {
    }
}

// tslint:disable-next-line:max-classes-per-file
export class ChangeLvl1 implements Action {
    type = 'change-lvl1';

    constructor(public payload: number) {
    }
}

// tslint:disable-next-line:max-classes-per-file
export class AutoUpdateLvl1 implements Action {
    type = 'auto-update';

    constructor() {
    }
}

export type DummyActions = ChangeLvl0 | ChangeLvl1 | AutoUpdateLvl1;

export interface DummyState {
    lvl0: {
        data: number,
        lvl1: {
            data: number,
        },
    };
}

export function* fibonacci() {
    let val1 = 1;
    let val2 = 1;
    let current;

    while (true) {
        current = val1;
        val1 = val2; // advance next value
        val2 = val2 + current;
        yield current;
    }
}
