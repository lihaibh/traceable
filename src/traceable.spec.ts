import 'reflect-metadata';
import 'rxjs/add/operator/skip';
import 'rxjs/add/operator/toPromise';
import 'should';

import { Effect } from '@lib/effect-decorator';
import { StateTraceable } from '@lib/state-traceable';
import * as _ from 'lodash';
import { Observable } from 'rxjs/Observable';
import { noop } from 'rxjs/util/noop';
import { spy, stub } from 'sinon';

import { Dummy, DummyActions, DummyState, first_state } from './_dummy';

/// <reference types="jest" />
// TODO: what suppose to happen when in runtime, using non Observable effect? should we throw on error?
// tslint:disable-next-line:no-big-function
describe('Traceable', () => {
    const _design_type = 'design:type';

    describe('State', () => {
        // TODO: create traceable object (using @Traceable or inherit from StateTraceable)
        it('should update state for a new manually performed action', async () => {
            // spy reduce to check number of calls to it
            const spyOn_reducer = spy(Dummy.prototype, 'reduce' as any);

            let dummy = new Dummy();

            // sanity check that
            spyOn_reducer.callCount.should.be.eql(0);
            expect(dummy.snapshot()).toMatchSnapshot();

            // Trigger the action that should change the state
            dummy.manualChangeLvl0(200);

            // Take the current state after performing the action
            // because the action is synchronized it should immediately change the initialized state
            // and that's what we will check, if the action, will update the state under the hood
            const observation = dummy.state$().take(1);

            // Use this hook to run the checks after state updates
            await observation.toPromise();
            expect(dummy.snapshot()).toMatchSnapshot();
            spyOn_reducer.callCount.should.be.eql(1);
            spyOn_reducer.restore();
            dummy.stop();
            dummy = null;
        });
        it('should spare calls to the reduce method for executing same action that did not modify the state', () => {
            // spy reduce to check number of calls to it
            const spyOn_reducer = spy(Dummy.prototype, 'reduce' as any);
            let dummy = new Dummy(_.set(_.cloneDeep(first_state), 'lvl0.data', 0));

            // sanity check
            spyOn_reducer.callCount.should.be.eql(0);

            // manually invoke action to change state's property to particular value and doing so multiple times
            for (let i = 0; i < 10; ++i) {
                dummy.manualChangeLvl0(20);
            }

            // first time because its a new action, second time because the the action changed the state
            spyOn_reducer.callCount.should.be.eql(2);
            spyOn_reducer.restore();
            dummy.stop();
            dummy = null;
        });
        it('should dispatch new state when it changes automatically by an action through an effect', async () => {
            const original_reducer = Dummy.prototype['reduce'];

            // spy reduce to check number of calls to it
            const stub_reducer = stub(Dummy.prototype, 'reduce' as any);
            stub_reducer.callsFake(function () {
                return original_reducer.apply(this, arguments);
            });

            let dummy = new Dummy(_.set(_.cloneDeep(first_state), 'lvl0.lvl1.data', 1000));

            // Sanity checks
            stub_reducer.callCount.should.be.eql(0);
            expect(dummy.snapshot()).toMatchSnapshot();

            // it should be updated with 9 fibonacci sequence numbers
            // 2 first values of fibonacci are identical, though the reduce may be executed many times
            // we should be notified only when state really changed
            // the state should be changed 9 times, we skip the first initial value
            const observation = dummy.state$().skip(1).take(9);

            observation.subscribe((state) => {
                expect(state).toMatchSnapshot();
            }, noop, () => {
                // check total numbers of calls to the reduce - including the trigger action
                // 1 by the trigger action + 10 by the side effect
                stub_reducer.callCount.should.be.eql(11);
                stub_reducer.restore();

                dummy.stop();
                dummy = null;
            });

            // this action should trigger a side effect that updates the state with fibonacci numbers
            dummy.triggerEffectForStateAutoUpdate();
        });
        it('should notify subscribers only when state content changes');
        // TODO: create two instances of the same class
        // TODO: execute action in one instance, expect 0 calls from the second and 1 call from the origin
        // TODO: check the same for different types
        it('should not mess up state of different prototype instances');

        describe('snapshot', () => {
            it('should have initialized snapshot by object creation', () => {
                const dummy = new Dummy(first_state);
                expect(dummy.snapshot()).toMatchSnapshot();
            });
            it('should point last state and receive it synchronously');
            it('should be able to select partial of the snapshot by function');
            it('should be able to select partial of the snapshot by path array');
            it('should be able to select partial of the snapshot by path arguments');
        });

        describe('state$', () => {
            it('should be able to track entire state changes');
            // TODO: change the state once, change the state in another place and check the number of subscription hits
            // TODO: check if only distinct changes are emitted
            it('should be able to select partial of the state by function and track it\'s changes');
            it('should be able to select partial of the state by path array and track it\'s changes');
            it('should be able to select partial of the state by path arguments and track it\'s changes');
        });

        describe('start', () => {
            it('should run effects that are only of Observable type in different inheritance hierarchies');
            it('should run effects only once regardless amount of calls');
        });

        describe('stop', () => {
            it('should disable effects from running in different inheritance hierarchies');
            it('should disable effects only once regardless amount of calls');
        });

        describe('reduce', () => {
            it('should emit new states when reduce modifies the state instance');
        });
    });

    describe('@Effect', () => {
        it('should throw an error for effects that are not of type Observable by type reflection', () => {

            class TraceableWithBadEffectType extends StateTraceable<DummyState, DummyActions> {
                // Decorate fields that are not exists yes - only created by constructing an object
                @Reflect.metadata(_design_type, Number) // Typescript should add this automatically
                private _bad_effect_1: number;
                @Reflect.metadata(_design_type, String) // Typescript should add this automatically
                private _bad_effect_2: string;
                @Reflect.metadata(_design_type, Object) // Typescript should add this automatically
                private _bad_effect_3: Object;
                @Reflect.metadata(_design_type, Observable) // Typescript should add this automatically
                private _good_effect_4: Observable<any> = Observable.from([1, 2, 3]);

                protected reduce(action: DummyActions, state: DummyState): DummyState {
                    return null;
                }

                protected initialize(): DummyState {
                    return null;
                }
            }

            const effectDecorator: PropertyDecorator = Effect();

            // Check runtime exceptions for property types

            expect(() => effectDecorator(TraceableWithBadEffectType.prototype, '_bad_effect_0'))
                .toThrowErrorMatchingSnapshot(); // property doesn't exists
            expect(() => effectDecorator(TraceableWithBadEffectType.prototype, '_bad_effect_1'))
                .toThrowErrorMatchingSnapshot();
            expect(() => effectDecorator(TraceableWithBadEffectType.prototype, '_bad_effect_2'))
                .toThrowErrorMatchingSnapshot();
            expect(() => effectDecorator(TraceableWithBadEffectType.prototype, '_bad_effect_3'))
                .toThrowErrorMatchingSnapshot();
            expect(() => effectDecorator(TraceableWithBadEffectType.prototype, '_good_effect_4'))
                .not.toThrow();

            // TODO: check types of static effects
        });
        it('should throw an error for effects that are not of type Observable by value', () => {
            // tslint:disable-next-line:max-classes-per-file
            class TraceableWithBadEffectValue extends StateTraceable<DummyState, DummyActions> {
                public _bad_effect_1;
                public _bad_effect_2;
                public _bad_effect_3;
                public _good_effect_1;
                // private static _bad_static_effect_1 = 5;
                // public static _good_static_effect_1: Observable<any> = Observable.from([1, 2, 3]);

                protected reduce(action: DummyActions, state: DummyState): DummyState {
                    return null;
                }

                protected initialize(): DummyState {
                    return null;
                }
            }

            // Test by property value - we can test by value on design time for prototype properties
            // because when the decorator runs it does not instantiate a new object, rather it
            // reads metadata about fields and and properties from the prototype object
            TraceableWithBadEffectValue.prototype._bad_effect_1 = 1;
            TraceableWithBadEffectValue.prototype._bad_effect_2 = 'another-bad-effect';
            TraceableWithBadEffectValue.prototype._bad_effect_3 = { bad: 'effect' };
            TraceableWithBadEffectValue.prototype._good_effect_1 = Observable.from([1, 2]).take(1);

            const effectDecorate: PropertyDecorator = Effect();

            // Check runtime exceptions for prototype property values
            expect(() => effectDecorate(TraceableWithBadEffectValue.prototype, '_bad_effect_0'))
                .toThrowErrorMatchingSnapshot(); // doesn't exists
            expect(() => effectDecorate(TraceableWithBadEffectValue.prototype, '_bad_effect_1'))
                .toThrowErrorMatchingSnapshot();
            expect(() => effectDecorate(TraceableWithBadEffectValue.prototype, '_bad_effect_2'))
                .toThrowErrorMatchingSnapshot();
            expect(() => effectDecorate(TraceableWithBadEffectValue.prototype, '_bad_effect_3'))
                .toThrowErrorMatchingSnapshot();
            expect(() => effectDecorate(TraceableWithBadEffectValue.prototype, '_good_effect_1'))
                .not.toThrow();

            // TODO: Check static fields
        });
        it('should throw an error if not being used in a traceable prototype' +
            ' (inherit from StateTraceable or using @Traceable decorator)', () => {
                // Load modules to test
                const _effect_module = require('./effect-decorator');
                const _traceable_module = require('./traceable-decorator');

                // Save original instances
                const originalEffectFactory = _effect_module.Effect;
                const originalTraceableFactory = _traceable_module.Traceable;
                const originalEffectDecorator = originalEffectFactory();
                const originalTraceableDecorator = originalTraceableFactory({
                    reduce: noop,
                    initialize: noop,
                });

                // Mock and spy functions
                const effectTester = stub(_effect_module, 'Effect');
                effectTester.callsFake(function () {
                    const _original_property_decorator =
                        originalEffectFactory.apply(this, arguments);

                    return function (): PropertyDecorator {
                        return _original_property_decorator.apply(this, arguments);
                    };
                });

                // Declare a non traceable and a traceable prototypes the JavaScript way
                // non traceable
                function Foo() {
                    this._effect_type = Observable.from([1, 2, 3]);
                }

                Foo.prototype._effect_value = Observable.from([1, 2, 3]);

                Reflect.decorate([
                    Reflect.metadata(_design_type, Observable),
                ], Foo.prototype, '_effect_type');

                // Inherits from StateTraceable
                function TraceableBar() {
                    this._effect_type = Observable.from([1, 2, 3]);
                    StateTraceable.call(this);
                }

                TraceableBar.prototype = _.create(StateTraceable.prototype, {
                    constructor: TraceableBar,
                });

                TraceableBar.prototype._effect_value = Observable.from([1, 2]);

                // We add metadata so we will be able to access the type of the
                // property before we construct an object
                Reflect.decorate([
                    Reflect.metadata('design:type', Observable),
                ], TraceableBar.prototype, '_effect_type');

                // Make Foo traceable by using the @Traceable decorator
                const TraceableFoo = originalTraceableDecorator(Foo);

                // Even if we created a TraceableFoo, Foo prototype should not be traceable
                // @Traceable decorator should not change Foo prototype itself
                expect(() => originalEffectDecorator(Foo.prototype, '_effect_type'))
                    .toThrowErrorMatchingSnapshot();
                expect(() => originalEffectDecorator(Foo.prototype, '_effect_value'))
                    .toThrowErrorMatchingSnapshot();

                // TODO: save content of Foo prototype and check its content after using the Traceab

                // Validate traceable prototype by using the @Traceable decorator
                expect(() => originalEffectDecorator(TraceableFoo.prototype, '_effect_type'))
                    .not.toThrow();
                expect(() => originalEffectDecorator(TraceableFoo.prototype, '_effect_value'))
                    .not.toThrow();

                // Validate traceable prototype by using the 'StateTraceable' class inheritance
                expect(() => originalEffectDecorator(TraceableBar.prototype, '_effect_type'))
                    .not.toThrow();
                expect(() => originalEffectDecorator(TraceableBar.prototype, '_effect_value'))
                    .not.toThrow();

                // TODO: Check the TypeScript way
                // TODO: Check static types
            });
    });

    describe('@Traceable', () => {
        it('should make the decorated class as "StateTraceable"');
        it('should make the decorated abstract class as "StateTraceable"');
        it('should throw an error when trying to use it in deeper inheritance level than the decorated class');
        it('should throw an error when trying to use in "StateTraceable" class');
        it('should throw an error when reduce function is not exists');
        it('should throw an error when initialize function is not exists');
        it('should throw an error for invalid reduce function type');
        it('should throw an error for invalid initialize function type');
    });

    describe('START_OWN_EFFECTS', () => {
        // tslint:disable-next-line:no-duplicate-string
        it('should throw an error when not being used on a traceable type');
        it('should invoke static effects only once for a particular traceable type, ' +
            'regardless inheritance hierarchy, when being invoked more than once');
    });

    describe('START_ALL_EFFECTS', () => {
        it('should throw an error when not being used on a traceable type');
        it('should start static effects only once for each class in the inheritance hierarchy of a traceable type, ' +
            'when being invoked more than once');
        it('should not start static effects for classes in different inheritance hierarchy of the traceable type');
    });

    describe('STOP_OWN_EFFECTS', () => {
        it('should throw an error when not being used on a traceable type');
        it('should stop static effects only once for a particular traceable type, ' +
            'regardless inheritance hierarchy, when being invoked more than once');
    });

    describe('STOP_ALL_EFFECTS', () => {
        it('should throw an error when not being used on a traceable type');
        it('should stop static effects only once for each class in the inheritance hierarchy of a traceable type, ' +
            'when being invoked more than once');
        it('should not stop static effects for classes in different inheritance hierarchy of the traceable type');
    });
});
