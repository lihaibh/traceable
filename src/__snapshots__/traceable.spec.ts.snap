// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Traceable @Effect should throw an error for effects that are not of type Observable by type reflection 1`] = `"Side effect TraceableWithBadEffectType._bad_effect_0 is not an Observable"`;

exports[`Traceable @Effect should throw an error for effects that are not of type Observable by type reflection 2`] = `"Side effect TraceableWithBadEffectType._bad_effect_1 is not an Observable"`;

exports[`Traceable @Effect should throw an error for effects that are not of type Observable by type reflection 3`] = `"Side effect TraceableWithBadEffectType._bad_effect_2 is not an Observable"`;

exports[`Traceable @Effect should throw an error for effects that are not of type Observable by type reflection 4`] = `"Side effect TraceableWithBadEffectType._bad_effect_3 is not an Observable"`;

exports[`Traceable @Effect should throw an error for effects that are not of type Observable by value 1`] = `"Side effect TraceableWithBadEffectValue._bad_effect_0 is not an Observable"`;

exports[`Traceable @Effect should throw an error for effects that are not of type Observable by value 2`] = `"Side effect TraceableWithBadEffectValue._bad_effect_1 is not an Observable"`;

exports[`Traceable @Effect should throw an error for effects that are not of type Observable by value 3`] = `"Side effect TraceableWithBadEffectValue._bad_effect_2 is not an Observable"`;

exports[`Traceable @Effect should throw an error for effects that are not of type Observable by value 4`] = `"Side effect TraceableWithBadEffectValue._bad_effect_3 is not an Observable"`;

exports[`Traceable @Effect should throw an error if not being used in a traceable prototype (inherit from StateTraceable or using @Traceable decorator) 1`] = `"You can't use @Effect decorator in a non traceable prototype"`;

exports[`Traceable @Effect should throw an error if not being used in a traceable prototype (inherit from StateTraceable or using @Traceable decorator) 2`] = `"You can't use @Effect decorator in a non traceable prototype"`;

exports[`Traceable State should dispatch new state when it changes automatically by an action through an effect 1`] = `
Object {
  "lvl0": Object {
    "data": 0,
    "lvl1": Object {
      "data": 1000,
    },
  },
}
`;

exports[`Traceable State should dispatch new state when it changes automatically by an action through an effect 2`] = `
Object {
  "lvl0": Object {
    "data": 0,
    "lvl1": Object {
      "data": 1,
    },
  },
}
`;

exports[`Traceable State should dispatch new state when it changes automatically by an action through an effect 3`] = `
Object {
  "lvl0": Object {
    "data": 0,
    "lvl1": Object {
      "data": 2,
    },
  },
}
`;

exports[`Traceable State should dispatch new state when it changes automatically by an action through an effect 4`] = `
Object {
  "lvl0": Object {
    "data": 0,
    "lvl1": Object {
      "data": 3,
    },
  },
}
`;

exports[`Traceable State should dispatch new state when it changes automatically by an action through an effect 5`] = `
Object {
  "lvl0": Object {
    "data": 0,
    "lvl1": Object {
      "data": 5,
    },
  },
}
`;

exports[`Traceable State should dispatch new state when it changes automatically by an action through an effect 6`] = `
Object {
  "lvl0": Object {
    "data": 0,
    "lvl1": Object {
      "data": 8,
    },
  },
}
`;

exports[`Traceable State should dispatch new state when it changes automatically by an action through an effect 7`] = `
Object {
  "lvl0": Object {
    "data": 0,
    "lvl1": Object {
      "data": 13,
    },
  },
}
`;

exports[`Traceable State should dispatch new state when it changes automatically by an action through an effect 8`] = `
Object {
  "lvl0": Object {
    "data": 0,
    "lvl1": Object {
      "data": 21,
    },
  },
}
`;

exports[`Traceable State should dispatch new state when it changes automatically by an action through an effect 9`] = `
Object {
  "lvl0": Object {
    "data": 0,
    "lvl1": Object {
      "data": 34,
    },
  },
}
`;

exports[`Traceable State should dispatch new state when it changes automatically by an action through an effect 10`] = `
Object {
  "lvl0": Object {
    "data": 0,
    "lvl1": Object {
      "data": 55,
    },
  },
}
`;

exports[`Traceable State should update state for a new manually performed action 1`] = `
Object {
  "lvl0": Object {
    "data": 0,
    "lvl1": Object {
      "data": 1,
    },
  },
}
`;

exports[`Traceable State should update state for a new manually performed action 2`] = `
Object {
  "lvl0": Object {
    "data": 200,
    "lvl1": Object {
      "data": 1,
    },
  },
}
`;

exports[`Traceable State snapshot should have initialized snapshot by object creation 1`] = `
Object {
  "lvl0": Object {
    "data": 0,
    "lvl1": Object {
      "data": 1,
    },
  },
}
`;
